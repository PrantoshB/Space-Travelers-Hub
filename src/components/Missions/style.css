table {
  margin: 2em 0;
  table-layout: fixed;
  width: 100%;
  text-align: left;
  border: 2px solid #e7e7e7;
  border-collapse: collapse;
}

th {
  padding: 0.4em;
}

thead th:nth-child(2) {
  width: 60%;
}

tbody tr:nth-child(odd) {
  background-color: #f3f0f0;
}

td {
  padding: 0.9em 0.4em;
  border: 1px solid #e7e7e7;
}

button {
  border: 1px solid #292828;
  border-radius: 5px;
  color: #292828;
  padding: 0.4em;
  text-transform: capitalize;
  cursor: pointer;
  width: 9rem;
}

.btn-active {
  color: #f83f3f;
  border-color: #f83f3f;
}
