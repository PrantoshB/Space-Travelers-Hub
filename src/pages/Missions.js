import React from 'react';
import MissionsTable from '../components/Missions';

const Missions = () => (
  <MissionsTable />
);

export default Missions;
