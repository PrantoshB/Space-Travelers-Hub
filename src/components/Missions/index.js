import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getMissionsAsync, joinMission } from '../../redux/missions/missions';
import Badge from '../Badge';
import './style.css';

const MissionsTable = () => {
  const dispatch = useDispatch();
  const data = useSelector((state) => state.missions);
  const { status, missions } = data;

  useEffect(() => {
    if (status === 'idle') {
      dispatch(getMissionsAsync());
    }
  }, [dispatch, status]);

  const handleClick = (mission) => {
    dispatch(joinMission(mission));
  };

  return (
    <table cellSpacing="0">
      <thead>
        <tr>
          <th scope="col">Mission</th>
          <th scope="col">Description</th>
          <th scope="col">Status</th>
          <th scope="col" aria-label="empty-th" />
        </tr>
      </thead>
      <tbody>
        {
          missions.map((mission) => {
            const {
              id, name, description, isReserved,
            } = mission;
            return (
              <tr key={id}>
                <td>{name}</td>
                <td>{description}</td>
                <td>
                  <Badge
                    text={isReserved ? 'Active member' : 'Not a member'}
                    active={isReserved}
                  />
                </td>
                <td>
                  <button
                    type="button"
                    className={`${isReserved && 'btn-active'}`}
                    onClick={() => handleClick(mission)}
                  >
                    {isReserved ? 'Leave Mission' : 'Join Mission'}
                  </button>
                </td>
              </tr>
            );
          })
        }
      </tbody>
    </table>
  );
};

export default MissionsTable;
