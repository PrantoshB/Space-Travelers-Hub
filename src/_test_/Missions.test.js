import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../pages/Missions';
import store from '../redux/configureStore';

it('matches snapshot', () => {
  const tree = renderer.create(
    <Provider store={store}><Missions /></Provider>,
  )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
