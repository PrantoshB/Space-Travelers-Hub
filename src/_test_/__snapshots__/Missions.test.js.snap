// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<table
  cellSpacing="0"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        Mission
      </th>
      <th
        scope="col"
      >
        Description
      </th>
      <th
        scope="col"
      >
        Status
      </th>
      <th
        aria-label="empty-th"
        scope="col"
      />
    </tr>
  </thead>
  <tbody />
</table>
`;
